### ISSUE TITLE
`reportUninitializedInstanceVariable` reports false positives in `@dataclass`

### ISSUE THREAD

**Issue by Tomaz-Vieira** [microsoft/pyright#3431]
**Describe the bug**
Using `reportUninitializedInstanceVariable` reports errors on fields that do not have a default value.

**To Reproduce**
```python
# pyright: strict, reportUninitializedInstanceVariable=error

from dataclasses import dataclass

@dataclass
class MyClass:
    x: int = 123 # no errors, as expected
    y: int # instance variable is not initialized in the class body or __init__ method
```

**Expected behavior**
No errors reported on `y: int`

**VS Code extension or command-line**
1.1.244

**Comment by erictraut** [comment]
I'm not able to repro the problem. Your example contains a bug that generates a different error unrelated to `reportUninitializedInstanceVariable`:

`Fields without default values cannot appear after fields with default values`

If you fix that bug, this type checks without errors.

**Comment by Tomaz-Vieira** [comment]
Sorry about the confusion, I added the `x` field to see if the error would occur also with a default value. Still, I get the error even if I remove my (admittedly poorly ordered) `x` field:

```python
# pyright: strict, reportUninitializedInstanceVariable=error

from dataclasses import dataclass

@dataclass
class MyClass:
    y: int 
```

the error disappears if I use `# pyright: strict, reportUninitializedInstanceVariable=none`

**Comment by erictraut** [comment]
I'm not seeing any error with the code sample above (once I remove the `x` field definition).

Pyright has special-case logic for dataclass to handle this case. I added this several months ago, but it looks like you're using the latest version.

Is it possible that you have some third-party stubs installed for the `dataclasses` library? Are you able to repro this problem with a clean Python environment?

**Comment by erictraut** [comment]
@Tomaz-Vieira, apologies for misdiagnosing the problem. I wasn't able to repro the problem earlier for reasons I don't understand. It must have been user error on my part — perhaps I was running an older version of pyright at the time.

In any case, I'm able to repro this now, and I've introduced a fix.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.245, which I just published. It will also be included in the next release of pylance.
